### PR TITLE
Remove unnecessary signature algorithms

### PR DESCRIPTION
**Pull request by coneiric**
---
**By submitting this pull-request, I confirm the following:**

- I have read and understood the contributor guide in [kovri-docs](https://github.com/monero-project/kovri-docs).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.
---

Remove signatures according to #498

### CODE REVIEW

File: src/core/router/info.h [hunk 1]
```diff
@@ -60,8 +60,8 @@ struct RouterInfoTraits
   /// @brief Router Info size constants
   enum Size : std::uint16_t
   {
-    MinBuffer = core::DSA_SIGNATURE_LENGTH,  // TODO(unassigned): see #498
-    MaxBuffer = 2048,  // TODO(anonimal): review if arbitrary
+    MinBuffer = core::EDDSA25519_SIGNATURE_LENGTH,  // TODO(unassigned): see #498
+    MaxBuffer = core::RSASHA5124096_KEY_LENGTH,  // TODO(anonimal): review if arbitrary
```
Comment (anonimal) [inline src/core/router/info.h]: >MaxBuffer = core::RSASHA5124096_KEY_LENGTH

This quickly breaks functionality. See the use-case for MaxBuffer.

Comment (coneiric) [reply src/core/router/info.h]: I didn't realize this breaks functionality. I'll revert MaxBuffer, and do some reading. Is the MinBuffer value still fine?

**Comment by coneiric** [comment]
I haven't checked with a full live test yet. Was able to start Kovri, and got it to connect to peers. Will test out if tunnel creation works.

Any tips for writing integration tests?

**Comment by anonimal** [comment]
>Any tips for writing integration tests for Kovri?

Aside from the prelim Boost.Python work and testnet, there's nothing currently in-place aside from running the router and testing tunnels manually.

**Comment by coneiric** [comment]
Just ran into a connection bug with a pull from the latest master. Thought it might have been something from my changes, so cloned a fresh copy, and same bug came up with building tunnels.

A few different errors keep popping up, (some rejections, timeouts, missing parts, etc.). The project builds fine, and runs without crashing.

Here is a gist of the [error logs](https://gist.github.com/coneiric/b3023dac2ad6d173342b35ec704c8a74)

Should I open an issue for this?

**Comment by anonimal** [comment]
>Just ran into a connection bug with a pull from the latest master

Those are all known and actual error reporting needs fine-tuning (tunnel building works regardless). Not related to this PR, no need for a new issue.

**Comment by anonimal** [comment]
```
@anonimal       oneiric: I'm doing a hefty amount of related review + related work on your PR. I'll end up committing something to your branch instead of sending diffs.
@anonimal       Essentially one of the problems is how we don't properly handle silent fails (somewhat silent) when encountering older routers in our netdb + when we want to build tunnels.
@anonimal       These older routers are signing with DSA-SHA1 and we never remove them from the netdb once parsed/created.
@anonimal       That and more. I've patched the other areas you missed but there's still more to do and I haven't caught it all yet.
@anonimal       This work would've been easier to finish if that identity rewrite was resolved first. Oh well, here we are.
@anonimal       s/how we don't/how we don't with this PR/
@anonimal       I've added much needed trace logging too, really makes things easier.
@anonimal       Dumping the RIs in hex and running kovri-util should be reduced to simply running GetDescription during runtime. I'll make the changes and more.
@anonimal       Anyway, next week (Monday).
```

The above reference was regarding nightmare code #366.

**Comment by anonimal** [comment]
An update/clarifications:

1. I've put this PR on the back-burner and will probably continue to do so for the rest of the month because its resolution is not a requisite for release. We'll see.
2. After reseed, most (but not all) of the time, we remove NetDb RI entries once they are confirmed DSA (set as unreachable, the NetDb eventually removes) but even when they are successfully removed, this isn't done quickly enough (saved ~1 minute later, deleted ~3/4 minutes after that) so, transports attempt to connect anyway - later failing during signature verification.
3. When both client / server tunnel Ed25519 keys are in use, and we attempt to use tunnels, there are silent failures that need to be logged and fixed
4. When server tunnel keys are DSA, obvious errors:
   - When HandlingDatabaseStoreMessage / creating remote LeaseSet from DatabaseStore, lease(s) is(are) DSA (keytype 0). We don't handle appropriately in LeaseSet's ReadFromBuffer (patched to catch, but not PR'd; if any lease is DSA, the entire leaseset is null - this is not a real solution), nor elsewhere (not patched).
     - Other non-Ed25519 behavior is TBD but it should be behave similar so long as the aforementioned isn't fixed.

**Comment by coneiric** [comment]
Thanks for your notes and work on this PR. Apologies for lack of reply, was unable to access my accounts.

> I've put this PR on the back-burner

Totally understand. I'll keep working on the points you brought up, hopefully ready by release after this one :)

> After reseed, most (but not all) of the time, we remove NetDb RI entries

Will look for areas in code to shorten the gap between confirming DSA reseed entries. Is there a minimum amount of time the entries need to stay in the NetDb?

> When both client / server tunnel Ed25519 keys are in use ... there are silent failures that need to be logged and fixed

Will look into sections of the code dealing with this scenario, and get an understanding of what those silent failures are.

> When HandlingDatabaseStoreMessage / creating remote LeaseSet from DatabaseStore, lease(s) is(are) DSA (keytype 0) ... if any lease is DSA, the entire leaseset is null - this is not a real solution)

Is it possible to separate the DSA entry from the rest of the LeaseSet? Does one member rely on all others in the set? Like in the catch scenario, was thinking about carving out the DSA entry(s), and replacing them with valid entries from elsewhere / doing a DatabaseStore reorg to only include ed25519 in LeaseSet. Not sure how that would work in code, or if it is even a good idea on the conceptual level.

Many thanks for the pointers on where to go from here.

**Comment by anonimal** [comment]
>Is it possible to separate the DSA entry from the rest of the LeaseSet?

IIRC, should be if implemented.

>Like in the catch scenario, was thinking about carving out the DSA entry(s), and replacing them with valid entries from elsewhere / doing a DatabaseStore reorg to only include ed25519 in LeaseSet.

Ok. Let's see what you come up with.

File: src/core/router/identity.cc [hunk 2]
```diff
@@ -435,60 +354,25 @@ SigningKeyType IdentityEx::GetSigningKeyType() const {
...
 
 CryptoKeyType IdentityEx::GetCryptoKeyType() const {
   if (m_StandardIdentity.certificate.type ==
       CERTIFICATE_TYPE_KEY && m_ExtendedBuffer)
     return bufbe16toh(m_ExtendedBuffer.get() + 2);  // crypto key
-  return CRYPTO_KEY_TYPE_ELGAMAL;
+  // TODO(unassigned): should an exception be raised here?
+  // since an unsupported setup is used when the above condition is false?
+  std::unique_ptr<std::uint8_t[]> null_key = std::make_unique<std::uint8_t[]>(m_ExtendedLen - 2);
+  return bufbe16toh(null_key.get());  // null crypto key
```
Comment (anonimal) [inline src/core/router/identity.cc]: >Likely not the correct solution, only fixes SIGILL runtime
interrupt.
>std::unique_ptr<std::uint8_t[]> null_key = std::make_unique<std::uint8_t[]>(m_ExtendedLen - 2);
>return bufbe16toh(null_key.get());  // null crypto key

Not a real fix. If you drop this commit and rebase the branch, I can start sending in the patches I mentioned earlier.

File: src/core/router/identity.cc [hunk 3]
```diff
@@ -435,60 +354,25 @@ SigningKeyType IdentityEx::GetSigningKeyType() const {
   if (m_StandardIdentity.certificate.type ==
       CERTIFICATE_TYPE_KEY && m_ExtendedBuffer)
     return bufbe16toh(m_ExtendedBuffer.get());  // signing key
-  return SIGNING_KEY_TYPE_DSA_SHA1;
+  // TODO(unassigned): should an exception be raised here?
+  // since an unsupported setup is used when the above condition is false?
+  std::unique_ptr<std::uint8_t[]> null_key = std::make_unique<std::uint8_t[]>(m_ExtendedLen);
+  return bufbe16toh(null_key.get());  // signing key
```
Comment (anonimal) [inline src/core/router/identity.cc]: >Likely not the correct solution, only fixes SIGILL runtime
interrupt.
>std::unique_ptr<std::uint8_t[]> null_key = std::make_unique<std::uint8_t[]>(m_ExtendedLen);
>return bufbe16toh(null_key.get());  // signing key

Not a real fix. If you drop this commit and rebase the branch, I can start sending in the patches I mentioned earlier.

Comment (coneiric) [reply src/core/router/identity.cc]: I can drop that commit no problem, where should I rebase the branch to? Just don't want all the comments to get deleted, unless that's not a problem.

**Comment by coneiric** [comment]
wtf, just tried to reset the last two commits, and Github auto-close the PR...

**Comment by anonimal** [comment]
Hmm, can you reopen?

**Comment by coneiric** [comment]
It says there are no commits. I'll try adding something simple, then reopen

**Comment by anonimal** [comment]
>Is there a minimum amount of time the entries need to stay in the NetDb?

@coneiric I missed this question. See `Time::RouterExpiration` and its implementation (both are in `src/core/router/net_db`). Also see the [docs](https://geti2p.net/en/docs/how/network-database) for details. Also note LeaseSet behavior and expiration time.

**Comment by anonimal** [comment]
#498 will be vastly easier to resolve once #366 and #643 and resolved.

**Comment by coneiric** [comment]
Closing until #366 and #643 are resolved.

**Pull request closed without merging**